OutOfSchool/OutOfSchool.WebApi/Services/StatisticService.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using OutOfSchool.Services.Models;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.WebApi.Extensions;
+using OutOfSchool.WebApi.Models;
+using Serilog;
+
+namespace OutOfSchool.WebApi.Services
+{
+    /// <summary>
+    /// Implements the operations to get popular workshops and categories.
+    /// </summary>
+    public class StatisticService : IStatisticService
+    {
+        private readonly IApplicationRepository applicationRepository;
+        private readonly IWorkshopRepository workshopRepository;
+        private readonly IEntityRepository<Category> categoryRepository;
+        private readonly ILogger logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StatisticService"/> class.
+        /// </summary>
+        /// <param name="applicationRepository">Application repository.</param>
+        /// <param name="workshopRepository">Workshop repository.</param>
+        /// <param name="categoryRepository">Category repository.</param>
+        /// <param name="logger">Logger.</param>
+        public StatisticService(
+            IApplicationRepository applicationRepository,
+            IWorkshopRepository workshopRepository,
+            IEntityRepository<Category> categoryRepository,
+            ILogger logger)
+        {
+            this.applicationRepository = applicationRepository;
+            this.workshopRepository = workshopRepository;
+            this.categoryRepository = categoryRepository;
+            this.logger = logger;
+        }
+
+        public async Task<IEnumerable<CategoryStatistic>> GetPopularCategories(int number)
+        {
+            logger.Information("Getting popular categories started.");
+
+            var categories = await categoryRepository.GetAll().ConfigureAwait(false);
```

Comment:
Check this example [enumerable vs queryable](https://dotnetcorecentral.com/blog/ienumerable-vs-iqueryable/). tl;dr version: our get all method returns every record in the database, so you load a complete list of categories from database every time and then start filtering/processing.
As you get alot of different entities from DB and do some processing with them - try to see if it's possible to make less requests to DB and load less data if you switch to iqueryable (modify repo to include iqueryable methods if needed) so that your filtering, ordering and counting happens in the database. The database queries are printed in console so you can see what's going on